src/libstore/build/derivation-goal.cc

```diff
@@ -124,6 +124,17 @@ DerivationGoal::DerivationGoal(const StorePath & drvPath, const BasicDerivation
     , buildMode(buildMode)
 {
     this->drv = std::make_unique<BasicDerivation>(BasicDerivation(drv));
+
+    auto outputHashes = staticOutputHashes(worker.store, drv);
```

Comment:
I'm a little  confused, is there any way we can abstract over this stuff in `haveDerivation` where it usual set? It looks like that is called just after?

Reply:
@regnat and I talked about this a rather meandering conversation (my fault!). @regnat mentioned the awkwardness of downcasting, but we do that a number of times already. `checkPathValidity` is the other place too.

Also, we should double check this is OK in the case of input-addressed remote buildings. Since the sender will strip away the `inputDrvs` but keep the output hashes, the remote side will calculate a different drvHash. Is this OK? I'm not sure.

Reply:
> @regnat mentioned the awkwardness of downcasting, but we do that a number of times already. checkPathValidity is the other place too.

I honestly don't have much of an opinion on this (except that we should eventually fix the root cause that forces us to downcast… but that's work for another day).

> Also, we should double check this is OK in the case of input-addressed remote buildings. Since the sender will strip away the inputDrvs but keep the output hashes, the remote side will calculate a different drvHash. Is this OK? I'm not sure.

I might obviously miss something, but it seems OK for me as this new `drvHash` is only used for registering the realisation remotely, meaning that
1. It doesn't change the output path of the derivation
2. It's never sent back to the sender

(and it's a correct output hash from the builder's pov, so at the very least it doesn't hurt to have it registered, although it's not strictly required for input-addressed drvs)